go/worker/keymanager/churp.go

```diff
@@ -0,0 +1,471 @@
+	)
+
+	w.submissions.Queue(status)
+}
+
+// submissionInfo contains details about a scheduled application submission.
+type submissionInfo struct {
+	// churpID represents the identifier of the CHURP scheme.
+	churpID uint8
+
+	// round indicates the round number for which an application is scheduled
+	// to be submitted.
+	round uint64
+
+	// epoch signifies the opening epoch for submissions.
+	epoch beacon.EpochTime
+
+	// height denotes the minimum block height for the submission.
+	height int64
+
+	// index specifies the position in the submission queue,
+	// or -1 if not queued.
+	index int
+
+	// cancel is a function to cancel the submission if it's in progress,
+	// otherwise, it's nil.
+	cancel context.CancelCauseFunc
+}
+
+// submissionScheduler is responsible for generating and submitting
+// application requests one epoch before handoffs.
+type submissionScheduler struct {
+	logger *logging.Logger
+
+	wg sync.WaitGroup
+
+	// kmWorker provides access to the common key manager services.
+	kmWorker *Worker
+
+	// queue contains submissions waiting to be processed, ordered by
+	// the minimum block height required for the submission.
+	queue SubmissionQueue
+
+	// running contains submissions that are currently in progress.
+	running map[uint8]*submissionInfo
+
+	// submissions contains both queued and running submissions.
+	submissions map[uint8]*submissionInfo
+}
+
+// newSubmissionScheduler creates a new submission scheduler.
+func newSubmissionScheduler(kmWorker *Worker) *submissionScheduler {
+	return &submissionScheduler{
+		logger:      logging.GetLogger("worker/keymanager/churp/submissions"),
+		kmWorker:    kmWorker,
+		queue:       make([]*submissionInfo, 0),
+		running:     make(map[uint8]*submissionInfo),
+		submissions: make(map[uint8]*submissionInfo),
+	}
+}
+
+// Queue schedules the given CHURP scheme for application submission,
+// updating or removing the scheduled submission if already present.
+func (s *submissionScheduler) Queue(status *churp.Status) {
+	removeFn := func(info *submissionInfo, cause error) {
+		if info.cancel != nil {
+			info.cancel(cause)
+			delete(s.running, status.ID)
+		}
+		if info.index != -1 {
+			heap.Remove(&s.queue, info.index)
+		}
+		delete(s.submissions, status.ID)
+	}
+
+	// Stop and remove submission for the previous round.
+	info, ok := s.submissions[status.ID]
+	if ok && info.round < status.Round {
+		removeFn(info, fmt.Errorf("new round"))
+	}
+
+	// Schedule submission for the current round.
+	info, ok = s.submissions[status.ID]
+	if !ok {
+		epoch := status.NextHandoff - 1
+		height, err := s.kmWorker.randomBlockHeight(epoch, 50)
+		if err != nil {
+			s.logger.Error("failed to select a random block height",
+				"err", err,
+			)
+			return
+		}
+
+		info = &submissionInfo{
+			churpID: status.ID,
+			round:   status.Round,
+			epoch:   epoch,
+			height:  height,
+			index:   -1,
+			cancel:  nil,
+		}
+
+		s.submissions[status.ID] = info
+		s.queue.Push(info)
```

Comment:
How does this insert into the right place given that queue should be sorted by height? Shouldn't this use `heap.Push`?

Reply:
Yes, fixed also for `Pop`.